pkg/scheduler/schedule_one.go

```diff
@@ -452,19 +452,22 @@ func (sched *Scheduler) findNodesThatPassFilters(
 	state *framework.CycleState,
 	pod *v1.Pod,
 	diagnosis framework.Diagnosis,
-	nodes []*framework.NodeInfo) ([]*v1.Node, error) {
-	numNodesToFind := sched.numFeasibleNodesToFind(int32(len(nodes)))
+	nodes []*framework.NodeInfo,
+	needUpdateNextStartNodeIndex bool) ([]*v1.Node, error) {
+	numAllNodes := len(nodes)
+	numNodesToFind := sched.numFeasibleNodesToFind(int32(numAllNodes))
 
 	// Create feasible list with enough space to avoid growing it
 	// and allow assigning.
 	feasibleNodes := make([]*v1.Node, numNodesToFind)
 
 	if !fwk.HasFilterPlugins() {
-		length := len(nodes)
 		for i := range feasibleNodes {
-			feasibleNodes[i] = nodes[(sched.nextStartNodeIndex+i)%length].Node()
+			feasibleNodes[i] = nodes[(sched.nextStartNodeIndex+i)%numAllNodes].Node()
+		}
+		if needUpdateNextStartNodeIndex {
+			sched.nextStartNodeIndex = (sched.nextStartNodeIndex + len(feasibleNodes)) % numAllNodes
 		}
-		sched.nextStartNodeIndex = (sched.nextStartNodeIndex + len(feasibleNodes)) % length
```

Comment:
if we move the update outside the function as suggested above, then we don't need to update `nextStartNodeIndex` here

Reply:
Done.